### PR TITLE
Fix #3846: Test for TextViewBindingAdapters

### LINKED ISSUE #3846: Add tests for TextViewBindingAdapters
TextViewBindingAdapters today doesn't have any tests for it, and ought to since it may not be obvious to track down future regressions that may occur in this file. These tests also ought to specifically test the binding adapters in their normal context (i.e. used through databinding with a test layout).

This also tracks adding new testing coverage after #3795 is merged since https://github.com/oppia/oppia-android/blob/baf7fd23affbf7dc27192a24fbb881e10519d2e7/utility/src/test/java/org/oppia/android/util/system/OppiaDateTimeFormatterTest.kt is being removed & its behaviors are now part of TextVIewBindingAdapters directly (so the same behaviors from that test suite should be added to the test suite for TextViewBindingAdapters).

### PR DESCRIPTION
**Pull request by rishidyno**
Fix #3846
## Essential Checklist

- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

### CODE REVIEW

**Comment by BenHenning** [comment]
Hi. As of today, some main reviewers have taken time off for the next few weeks, so it may take a little while before we can look at this PR. We appreciate your patience while some of our team members recharge. We'll be fully returning on 4 January 2021.

**Comment by rishidyno** [comment]
.

**Comment by rishidyno** [comment]
@BenHenning @rt4914 All tests have passed PTAL. 

https://user-images.githubusercontent.com/78549195/149935135-6bb790a8-cbd8-49fd-8904-7b36f1d89708.mp4

File: app/src/main/AndroidManifest.xml
Comment (BenHenning) [inline app/src/main/AndroidManifest.xml]: Is this in lexicographical order by android:name?

File: app/src/main/res/drawable/test_text_view_drawable_binding_adapter.xml
Comment (BenHenning) [inline app/src/main/res/drawable/test_text_view_drawable_binding_adapter.xml]: Could you maybe use one of the drawables already included in the app rather than adding a new one?

File: app/src/main/res/layout/test_text_view_bindable_adapter_activity.xml
Comment (BenHenning) [inline app/src/main/res/layout/test_text_view_bindable_adapter_activity.xml]: Nit: should be on the line above.

File: app/src/main/res/values/strings.xml
Comment (BenHenning) [inline app/src/main/res/values/strings.xml]: Nit: please make sure there's exactly one space on either side of the activity name.

File: app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt]: Please also remove the unnecessary KDocs in this test suite.

File: app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt]: Nit: remove blank newline for consistency.

File: app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt:83]: Similar to a comment I left on an earlier PR, this doesn't actually point to a class or object (so it shouldn't use square brackets).

File: app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt]: These tests cover a couple of the happy path cases, but ought to cover all other cases that the production code includes.

One way to do this: you can analyze the production code to determine different cases of input that would hit each branch of the implementation. Then, consider what behaviors would result in those inputs, and then add tests for each behavior you come up with. In this case, you've only add one behavior for setProfileDataText (setting it to the same timestamp as the current time).

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @rishidyno! Left some comments--PTAL.

File: app/src/main/AndroidManifest.xml
Comment (rishidyno) [reply app/src/main/AndroidManifest.xml]: Yup!

File: app/src/main/res/drawable/test_text_view_drawable_binding_adapter.xml
Comment (rishidyno) [reply app/src/main/res/drawable/test_text_view_drawable_binding_adapter.xml]: Ok!

File: app/src/main/res/layout/test_text_view_bindable_adapter_activity.xml
Comment (rishidyno) [reply app/src/main/res/layout/test_text_view_bindable_adapter_activity.xml]: Done

File: app/src/main/res/values/strings.xml
Comment (rishidyno) [reply app/src/main/res/values/strings.xml]: Done

File: app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt
Comment (rishidyno) [reply app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt:83]: Well, for this It's done this way in other binding adapters test files. Otherwise, makes it in consistence.

File: app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt
Comment (rishidyno) [reply app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt]: Done

File: app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt
Comment (rishidyno) [reply app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt]: Done.

File: app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt
Comment (rishidyno) [reply app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt:83]: So I have left it as is. Or will have to change others?

File: app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt
Comment (rishidyno) [reply app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt]: Regarding this, I quite not understand where to get the production code. Earlier when we were in the meet I used resource handler and got the current time in ms with Oppia clock and just used that function which converts it to that format and compared. Will it work if I do that.

**Comment by rishidyno** [comment]
@BenHenning PTAL.

**Comment by BenHenning** [comment]
@rishidyno is this ready for review? I don't see any of my comments responded to. Please make sure to respond to all comments to confirm they've been addressed, or why they haven't in cases when they can't or you don't think they should.

**Comment by rishidyno** [comment]
Reason for me not responding [here](https://github.com/oppia/oppia-android/pull/4077#issuecomment-1018264579) as mention in #4077 Pr.

**Comment by rishidyno** [comment]
> @rishidyno is this ready for review? I don't see any of my comments responded to. Please make sure to respond to all comments to confirm they've been addressed, or why they haven't in cases when they can't or you don't think they should.

Reason for me not responding [here](https://github.com/oppia/oppia-android/pull/4077#issuecomment-1018264579) as mention in #4077 Pr.

**Comment by rishidyno** [comment]
image
image
image

**Review by rishidyno — COMMENTED** [review]
@BenHenning PTAl

**Comment by rishidyno** [comment]
@BenHenning I think I did not submit it to get a review, that's why they were not showing. Thanks to @yash10019coder he helped me figure it out. PTAL.

File: app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt
Comment (BenHenning) [reply app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt:83]: We should fix it since it's incorrect. The other test suites ought to be updated, too, but that's outside the scope of this PR.

File: app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt
Comment (BenHenning) [reply app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt]: The production code for setProfileDataText is https://github.com/oppia/oppia-android/blob/b9d4b13af8c880f5fed8099194f26f9b3f52f0c3/app/src/main/java/org/oppia/android/app/databinding/TextViewBindingAdapters.java#L26. The way to think about this is any if/when blocks and loops indicate multiple conditions that should have corresponding tests. Sometimes methods without control structures can also vary based on input (such as selecting a different string). We should test all the cases we can think of for these methods.

File: app/src/main/AndroidManifest.xml
Comment (BenHenning) [reply app/src/main/AndroidManifest.xml]: Shouldn't it go before TopicTestActivity ('e' comes before 'o')?

**Review by BenHenning — COMMENTED** [review]
Thanks @rishidyno! Just a few comments left, including the one requesting to add more tests. PTAL.

File: app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt
Comment (rt4914) [reply app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt]: Basically `setProfileDataText ` is only responsible for showing "Created on ....." text in textview, which means we are just sending the timestamp as long and it is converting it as string. Now whatever timestamp we send it will convert it to string, that timestamp value can be `0` or `1556094120000` and it will just covert it to either 

`Created on 1 Jan 1970`, or `Created on April 24, 2019`. So I am not sure what other test we can add. 

Are you thinking of other locales like arabic?

**Comment by rt4914** [comment]
Will need to rely on @BenHenning for review on this one.

File: app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt
Comment (BenHenning) [reply app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt]: I think testing setProfileDataText for two different inputs is probably sufficient (to demonstrate that the variable part of the string is working as expected). We don't need to test multiple locales here since there's nothing locale-specific.

setProfileLastVisitedText has a lot more behaviors to test, though, since the label that's computed has a bunch of different potential cases. I see that method as being the one that will have the most tests corresponding to it.

File: app/src/main/AndroidManifest.xml
Comment (rishidyno) [reply app/src/main/AndroidManifest.xml]: Ya! My bad.

File: app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt
Comment (rishidyno) [reply app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt]: Added another case of `TIMESTAMP =0`.

**Comment by rishidyno** [comment]
@BenHenning PTAL.

File: app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt
Comment (rishidyno) [reply app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt]: So I think checking `setProfileLastVisitedText` for `just now`, `x minutes ago`, `x hours ago` and `x days ago` will be sufficient?

Comment (rishidyno) [reply app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt]: @BenHenning @rt4914 I have added other test cases for possible outputs. Do I need to structure them?

Comment (BenHenning) [reply app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt]: Thanks @rishidyno! I think this is on the right track.

In terms of testing coverage, I think you did a really good job covering all of the "this should work as expected" cases. The three that I think are also worth covering are when the provided timestamp is in the future (i.e. bigger than the current time), when it's zero, and when it's negative. These are all cases handled by the implementation, so we should verify that they work as expected (as per the adapter's contract).

Beyond that, you should structure each behavior being checked in its own test rather than combining them all into one.

**Review by BenHenning — COMMENTED** [review]
Thanks @rishidyno! Apologies for the delayed review. I think the PR is quite close, but I had some suggestions to improve upon the tests you recently added--PTAL.

File: app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt
Comment (rishidyno) [reply app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt]: @BenHenning I have separated all test cases and added few more. PTAL.

File: app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt]: Rather than case1, case2, etc. you should name this based on the conditions and outcome, per https://github.com/oppia/oppia-android/wiki/Oppia-Android-Testing#naming-convention.

For this specific method, it could be called:
``testSetProfileLastVisitedText_forTimeTwoDaysAgo_setsLastUsedTwoDaysAgoText``

Ditto elsewhere.

File: app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt]: Perhaps we could add a new string that just says "Last used recently" that we update the adapter to use in these cases, instead. It's nice that we have tests since they demonstrate that the resulting behavior isn't ideal. I know this is a bit outside the scope of the PR, but it seems like a nice thing to fix while someone's looking at this code, anyway. WDYT?

File: app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt
Comment (BenHenning) [reply app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt]: Left some follow-up comments, so resolving this now.

**Review by BenHenning — COMMENTED** [review]
Thanks @rishidyno! Just had a few follow-up comments--PTAL.

File: app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt
Comment (rishidyno) [reply app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt]: I think doing this would be easy, as Instead of returning an empty string in these cases, we can just return `recently`.

File: app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt
Comment (rishidyno) [inline app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt:252]: @BenHenning Is this fun name suitable?

File: app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt
Comment (rishidyno) [reply app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt]: Done.

File: app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt
Comment (rishidyno) [reply app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt]: I have already made few changes PTAL.

Comment (rishidyno) [reply app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt]: I used google translator for `'recently'` for other languages.

**Comment by BenHenning** [comment]
Apologies; will  need to look at this tomorrow.

File: app/src/main/res/values-ar/strings.xml
Comment (BenHenning) [inline app/src/main/res/values-ar/strings.xml]: Prefer not to add this or the Brazilian Portuguese variant--we'll have translators translate these after the English base string is checked in.

File: app/src/main/res/values/strings.xml
Comment (BenHenning) [inline app/src/main/res/values/strings.xml]: The name ought to be clearer (the other names aren't very clear). Perhaps: ``last_logged_in_recently`` would be clearer?

File: app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt]: Nit: ``TwoHoursAgo``. Ditto elsewhere (I think there are some pluralization issues in other tests).

File: app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt
Comment (BenHenning) [reply app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt:252]: Yup, this is clear--thanks for checking!

File: app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt
Comment (BenHenning) [reply app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt]: Followed up with additional comments--thanks!

**Review by BenHenning — COMMENTED** [review]
Thanks @rishidyno! I think the latest changes look really good! Just had a few nits--PTAL.

File: app/src/main/res/values-ar/strings.xml
Comment (rishidyno) [reply app/src/main/res/values-ar/strings.xml]: Removed.

File: app/src/main/res/values/strings.xml
Comment (rishidyno) [reply app/src/main/res/values/strings.xml]: Changed.

File: app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt
Comment (rishidyno) [reply app/src/sharedTest/java/org/oppia/android/app/databinding/TextViewBindingAdaptersTest.kt]: Done.

**Comment by rishidyno** [comment]
@BenHenning did few nit-fix PTAL.

**Review by BenHenning — APPROVED** [review]
Thanks @rishidyno. This LGTM!

**Comment by BenHenning** [comment]
@anandwana001 and @rt4914 I think you need to take a pass for codeowners.

**Review by anandwana001 — APPROVED** [review]
My code owner file LGTM

**Review by rt4914 — APPROVED** [review]
Thanks a lot @rishidyno 
LGTM

**Pull request merged**
